pdf/pdfrenderthread.cpp

```diff
@@ -126,7 +127,28 @@ class PDFRenderThreadPrivate
             delete page;
         }
     }
-
+    void retrieveTextBoxes(int i)
+    {
+        if (i < 0 || i >= document->numPages()) {
+            return;
+        }
+        Poppler::Page* page = document->page(i);
+        QList<Poppler::TextBox*> plst = page->textList();
+        QList< QPair< QRectF, QString > > lst;
+        for (QList<Poppler::TextBox*>::const_iterator box = plst.begin();
+             box != plst.end(); box++) {
+            QSizeF psize = page->pageSize();
+            QRectF bbox = (*box)->boundingBox();
+            QRectF rect{ bbox.x() / psize.width(),
+                    bbox.y() / psize.height(),
+                    bbox.width() / psize.width(),
+                    bbox.height() / psize.height()};
+            lst.append(QPair<QRectF, QString>{rect, (*box)->text() + ((*box)->hasSpaceAfter() ? " " : "")});
```

Comment:
I've switched to use the Poppler structure  itself. I agree there is no point in reimplementing all Poppler objects. The only issue that makes me kept the QPair is that the bounding box is given in 1/72 of inches which requires to have the page size at hand always and do conversion each time we use the box. So I store a QPair<QRectF, Poppler::TextBox*>, instead of the previous QPair<QRectF, QString>.

Then space is added when mentioned and not for last word in selection.